File: projects/multi_dashboard/frontends/src/api/finance.ts
import axios from "axios";

const FINANCE_API_BASE = (
  import.meta.env.VITE_FINANCE_API_BASE ?? "/finance-api"
).replace(/\/$/, "");
const client = axios.create({ baseURL: FINANCE_API_BASE });

type Kind = "investment" | "mortgage" | "salary_projection";

export type InvestmentParams = {
  current_balance: number;
  monthly_payment: number;
  cagr_percent: number;
  years: number;
  contribution_timing: "start" | "end";
  lumpsums?: { month: number; amount: number }[];
};

export type MortgageParams = {
  principal: number;
  annual_rate: number;
  years: number;
  monthly_payment?: number;
  extra_monthly?: number;
  extra_annual?: number;
  annual_lump_month?: number;
};

export type SalaryProjectionParams = {
  monthly_salary: number;
  yearly_increment_percent: number;
  monthly_expense: number;
  yearly_inflation_percent: number;
  years: number;
  annual_return_percent: number;
};

export async function financeCompute<
  TParams extends InvestmentParams | MortgageParams | SalaryProjectionParams,
>(kind: Kind, params: TParams) {
  const { data } = await client.post<{ ok: boolean; result: any }>("/compute", {
    kind,
    params,
  });
  return data.result;
}

File: projects/multi_dashboard/backends/finance-py/app/schemas/finance.py
from __future__ import annotations

from typing import Annotated, Literal, Optional, Sequence

from pydantic import BaseModel, Field


class Lump(BaseModel):
    month: Annotated[int, Field(ge=0)]
    amount: Annotated[float, Field(ge=0)]


class InvestmentParams(BaseModel):
    current_balance: Annotated[float, Field(ge=0)] = 0.0
    monthly_payment: Annotated[float, Field(ge=0)] = 0.0
    cagr_percent: Annotated[float, Field(ge=0)] = 0.0
    years: Annotated[int, Field(ge=0)] = 0
    contribution_timing: Literal["start", "end"] = "end"
    lumpsums: Optional[Sequence[Lump]] = None


class MortgageParams(BaseModel):
    principal: Annotated[float, Field(gt=0)]
    annual_rate: Annotated[float, Field(ge=0)]
    years: Annotated[int, Field(gt=0)]
    monthly_payment: Optional[Annotated[float, Field(ge=0)]] = None
    extra_monthly: Annotated[float, Field(ge=0)] = 0.0
    extra_annual: Annotated[float, Field(ge=0)] = 0.0
    annual_lump_month: Annotated[int, Field(ge=1, le=12)] = 12


class SalaryProjectionParams(BaseModel):
    monthly_salary: Annotated[float, Field(gt=0)]
    yearly_increment_percent: Annotated[float, Field(ge=0)] = 3.0
    monthly_expense: Annotated[float, Field(ge=0)]
    yearly_inflation_percent: Annotated[float, Field(ge=0)] = 3.0
    years: Annotated[int, Field(ge=1, le=100)]
    annual_return_percent: Annotated[float, Field(ge=0)] = 18.0


Kind = Literal["investment", "mortgage", "salary_projection"]


class ComputeRequest(BaseModel):
    kind: Kind
    params: dict


class ComputeResponse(BaseModel):
    ok: bool
    result: dict
